File: lecture11/serverSide1_hints_kadai3.js
document.addEventListener(
  "DOMContentLoaded",
  function () {
    // 各所にイベントリスナーを設置する．これは，HTML読み込み完了時に動作する．

    class webConnector extends XMLHttpRequest {
      constructor() {
        super();
        this.addEventListener(
          "loadstart",
          function () {
            window.alert(this.url + "と通信を開始しました");
          },
          false
        ); //通信開始時の処理
        this.addEventListener(
          "load",
          function () {
            let data = this.response;

            // 課題3-3で下記構造のリストを作る関数show_kadai3
            // data: [
            //     {'name': name, 'message': message},
            //     {'name': name, 'message': message},
            //     {'name': name, 'message': message},
            // ]
            show_kadai3(data);
          },
          false
        ); //通信成功時処理
        this.addEventListener(
          "error",
          function () {
            window.alert(this.url + "との通信に失敗しました");
          },
          false
        ); //通信失敗時処理
      }
    }
    let xhr = new webConnector();

    // 課題2-2で作成，3-3でも使う
    document.getElementById("kadai2").addEventListener(
      "click",
      function () {
        let name = document.getElementById("name").value;
        let message = document.getElementById("message").value;

        let url =
          "./cgi-bin/serverSide1.py?" + "&name=" + name + "&message=" + message;
        xhr.open("GET", url);
        xhr.responseType = "json";
        xhr.url = url;
        xhr.send(null);
      },
      false
    );
  },
  false
);

// 課題3-3で作成
function show_kadai3(data) {
  // idがkadai3のタグの取得
  let kadai3 = document.getElementById("kadai3");

  if (data === null) {
    kadai3.textContent = "メッセージはありません";
  } else {
    // 今回は表示用にデータを下記のように整形してみる
    // 表示イメージ
    // ● name
    //      〇 「message」
    // ● name
    //      〇 「message」
    // ● name
    //      〇 「message」
    let text = "<ul>";

    // data内構造は下記のようになっている．
    // [{'name': name, 'message': message},{'name': name, 'message': message},{'name': name, 'message': message}]
    for (let dictionary of data) {
      text +=
        "<li>" +
        dictionary["name"] +
        "<ul><li>「" +
        dictionary["message"] +
        "」</li></ul></li>";
    }
    text += "</ul>";

    // htmlへ反映（現状何回送信しても表示されるのはダミーデータと直近の送信データのみ）
    kadai3.innerHTML = text;
  }
}

File: lecture11/serverSide1_hints_kadai3.py
#!/usr/bin/python3

import cgi
import cgitb
import os
import sys
import json

if __name__ == "__main__":
    if "REQUEST_METHOD" in os.environ:
        param = cgi.FieldStorage()  # Formからやってきたデータを受け取る

        # 課題2-3で作成,3-1でも使う
    name = param.getfirst("name", 0)
    message = param.getfirst("message", 0)

    if name != 0 and message != 0:

        # 課題3-1で作成
        # 送信用のリストsend_listを作成
        send_list = []

        # リストに辞書を追加
        # appendメソッドを使うと，htmlからの入力が，name = '田吾作’, message = ’耕すぞー’だったとしたら，
        # send_listは，[{‘name’:’田吾作’,’message’:’耕すぞー’}]となる．
        send_list.append({"name": name, "message": message})

        # 赤いきつねを追加する．nameに「きつね」を，messageに「赤！」を直接書き込む．
        # send_listは，[{‘name’:’田吾作’,’message’:’耕すぞー’}, {‘name’:’きつね’,’message’:’赤！’}]となる．
        send_list.append({"name": "きつね", "message": "赤！"})

        # 赤いきつねと同じように緑のたぬきを追加する．
        # send_listは，[{‘name’:’田吾作’,’message’:’耕すぞー’}, {‘name’:’きつね’,’message’:’赤！’}, {‘name’:’たぬき’,’message’:’緑！’}]となる
        send_list.append({"name": "たぬき", "message": "緑！"})

        print("Content-Type: application/json\n\n")

        print(
            json.dumps(
                # 返送するのは作成したリストsend_list．中身は [{‘name’:’田吾作’,’message’:’耕すぞー’}, {‘name’:’きつね’,’message’:’赤！’}, {‘name’:’たぬき’,’message’:’緑！’}]である．
                send_list
            )
        )
